### DISCUSSION TITLE
Fail to install 3.0.5 on M1 Mac 12.3.1:   psych

### DISCUSSION THREAD

**Discussion by jrochkind** [rbenv/ruby-build#2101] [Build failures] [answered]
I am not totally certain if I've succesfully installed earlier versions of 3.0.x on this machine with ruby-build or not. But I can't currently install 3.0.5. 

Is it _expected_ that should be able to install on an M1 Mac with homebrew, using ruby-build, without custom configuration for library locations and such? Or is custom configuration for library locations expected to be needed? 

Here is my output:

```
ruby-build 3.0.5 ~/.rubies/ruby-3.0.5
To follow progress, use 'tail -f /var/folders/_1/89lqv5550mx2tggl22z27_p18516fz/T/ruby-build.20221128171707.72665.log' or pass --verbose
Downloading openssl-1.1.1s.tar.gz...
-> https://dqw8nmjcqpjn7.cloudfront.net/c5ac01e760ee6ff0dab61d6b2bbd30146724d063eb322180c6f18a6f74e4b6aa
Installing openssl-1.1.1s...
Installed openssl-1.1.1s to /Users/jrochkind/.rubies/ruby-3.0.5

Downloading ruby-3.0.5.tar.gz...
-> https://cache.ruby-lang.org/pub/ruby/3.0/ruby-3.0.5.tar.gz
Installing ruby-3.0.5...
ruby-build: using readline from homebrew
ruby-build: using gmp from homebrew

BUILD FAILED (macOS 12.3.1 using ruby-build 20221124)
```

Later:

```
installing bundled gem cache:       /Users/jrochkind/.rubies/ruby-3.0.5/lib/ruby/gems/3.0.0/cache
The Ruby openssl extension was not compiled.
/Users/jrochkind/.rubies/ruby-3.0.5/lib/ruby/3.0.0/yaml.rb:3: warning: It seems your ruby installation is missing psych (for YAML output).
To eliminate this warning, please install libyaml and reinstall your ruby.
The Ruby yaml extension was not compiled.
ERROR: Ruby install aborted due to missing extensions
Configure options used:
  --prefix=/Users/jrochkind/.rubies/ruby-3.0.5
  --with-openssl-dir=/Users/jrochkind/.rubies/ruby-3.0.5/openssl
  --enable-shared
  --with-readline-dir=/opt/homebrew/opt/readline
  --with-gmp-dir=/opt/homebrew/opt/gmp
  CC=clang
  LDFLAGS=-L/Users/jrochkind/.rubies/ruby-3.0.5/lib
  CPPFLAGS=-I/Users/jrochkind/.rubies/ruby-3.0.5/include
```

Things that seem weird to me:

* It says "The Ruby openssl extension was not compiled.", but earlier output seemed to show openssl getting installed fine? I'm not sure if this is relevant to the problem. 

* The error message seems to say it's having trouble installing psych, but earlier output made it look like it already had installed psych:

      ```
      installing default gems from ext:   /Users/jrochkind/.rubies/ruby-3.0.5/lib/ruby/gems/3.0.0
      [...]
                                            psych 3.3.2
      [...]
      ```

I do have libyaml installed, with homebrew. 

```
$ brew list libyaml
/opt/homebrew/Cellar/libyaml/0.2.5/include/yaml.h
/opt/homebrew/Cellar/libyaml/0.2.5/lib/libyaml-0.2.dylib
/opt/homebrew/Cellar/libyaml/0.2.5/lib/pkgconfig/yaml-0.1.pc
/opt/homebrew/Cellar/libyaml/0.2.5/lib/ (2 other files)
```

But maybe the ruby compilation step can't find it? Do I need to supply custom build configuration? Is it expected that I would need to?  

In the past I have installed ruby's with `ruby-build` without needing to do any custom configuration, it just worked. Not sure if something has changed in my environment, in ruby-build, in the newest version of ruby, or what. 

Thanks for any advice!

**Comment by mislav** [comment, marked as answer]
Thanks for reporting! I'm sorry you're having troubles.

> Is it _expected_ that should be able to install on an M1 Mac with homebrew, using ruby-build, without custom configuration for library locations and such? Or is custom configuration for library locations expected to be needed?

I expect it to work. Otherwise, ruby-build is broken. However, when people encounter build failures due to edge-cases (or hard-to-debug scenarios) specific to their particular environment, we often suggest passing custom configuration flags to help the build to succeed.

> But maybe the ruby compilation step can't find it? Do I need to supply custom build configuration? Is it expected that I would need to?

You can try:
```sh
brew install openssl@1.1 libyaml
RUBY_CONFIGURE_OPTS="--with-openssl-dir=$(brew --prefix openssl@1.1) --disable-install-doc" ruby-build 3.0.5 ~/.rubies/ruby-3.0.5
```

Explanation:
- ruby-build will link to Homebrew's libyaml automatically.
- This links to Homebrew's openssl, which isn't strictly required, but it will speed up your compilation because ruby-build won't have to compile `openssl-1.1.1s` anymore. It might get you over the `The Ruby openssl extension was not compiled` failure, which I have to admit that I don't know the cause of without full build logs.
- This skips building RI and rdocs, which speeds up your build attempts.

**Reply by jrochkind** [reply]
Thanks! I don't have a preference to use homebrew libyaml if there's another default more standard option, that was just me trying to troubleshoot!

But since libyaml seemed to be my problem... in the interest of trying the smallest intervention that might work, I first tried without the homebrew openssl link, to see if it would work without that:

    RUBY_CONFIGURE_OPTS="--with-libyaml-dir=$(brew --prefix libyaml) --disable-install-doc" ruby-build 3.0.5 ~/test-ruby-3.0.5

That worked! 

the openssl thing was a red herring -- I think maybe it was a bug that the log output said something about "The Ruby openssl extension was not compiled", when in fact there were no problems there. 

Thanks! Glad to see the brew libyaml situation is already improved in a PR!  (I am curious why this became necessary when it did not used to be, or what other libyaml options exist on MacOS other than brew!)

**Reply by stephannv** [reply]
Thanks @jrochkind this solved my problem.
The error was:
```
warning: It seems your ruby installation is missing psych (for YAML output).
To eliminate this warning, please install libyaml and reinstall your ruby.
```

Then I ran:
```
RUBY_CONFIGURE_OPTS="--with-libyaml-dir=$(brew --prefix libyaml) --disable-install-doc" asdf install ruby 3.2.0
```

**Comment by jegacs** [comment]
Hello! I have tried this very same commands and I have not been able to make my local setup work. I'm seeing the very same error @jrochkind experienced before he could fix his installation. All the versions are the same actually, I can install up to ruby 3.0.3 but not 3.0.4 nor 3.0.5. 
Do you happen to have any other recommendation to fix this? Thank you.

**Reply by jegacs** [reply]
In case is helpful for anyone, what I did to do the trick was to append the --disable-shared flag to the compiler options. That made the trick for me.

**Comment by yonkeltron** [comment]
I can confirm that this issue is occurring with Ruby 3.2.1 on macOS Monterey where libyaml is installed with Homebrew. Specifically I ran `ruby-build 3.2.1 ~/tmp` from my home directory and it failed to complete. 

Here's the relevant section from the build log:

```
installing bundled gem cache:       /Users/USERNAME/tmp/lib/ruby/gems/3.2.0/cache
Ignoring psych-5.1.0 because its extensions are not built. Try: gem pristine psych --version 5.1.0
Ignoring stringio-3.0.5 because its extensions are not built. Try: gem pristine stringio --version 3.0.5
/Users/USERNAME/tmp/lib/ruby/3.2.0/yaml.rb:3: warning: It seems your ruby installation is missing psych (for YAML output).
To eliminate this warning, please install libyaml and reinstall your ruby.
The Ruby yaml extension was not compiled.
ERROR: Ruby install aborted due to missing extensions
Configure options used:
  --prefix=/Users/USERNAME/tmp
  --with-openssl-dir=/Users/USERNAME/tmp/openssl
  --with-readline-dir=/usr/local/opt/readline
  --with-gmp-dir=/usr/local/opt/gmp
  --enable-yjit
  --with-libyaml-dir=/usr/local/opt/libyaml
  --disable-shared
  --disable-install-doc
  CC=clang
  LDFLAGS=-L/Users/USERNAME/tmp/lib
  CPPFLAGS=-I/Users/USERNAME/tmp/include
```

Are there any other bits of information I can provide to help?

**Reply by mislav** [reply]
No ideas, but someone else is reporting a similar problem too: https://github.com/rbenv/ruby-build/discussions/2012#discussioncomment-5444313

1. Do you happen to have any `GEM*` environment variables at the time of compilation?
2. Can you directly invoke `bin/ruby` in the directory of the failed Ruby build and try loading psych manually?

**Reply by yonkeltron** [reply]
@mislav I am checking right now on both of these things and will see if I can write a script which can reproduce the exact issue. Do you want the output in this thread or in the issue?

**Reply by mislav** [reply]
@yonkeltron Please start a new thread, with the title describing that psych was built successfully but that the compilation process was still aborted with what looks like an invalid error.

Also please include information about whether the next Ruby version (`3.0.5`) also exhibits the same problem.

**Reply by yonkeltron** [reply]
Done! https://github.com/rbenv/ruby-build/discussions/2170

**Comment by initseis** [comment]
This tutorial https://velog.io/@inhalin/warning-It-seems-your-ruby-installation-is-missing-psych-for-YAML-output by @inhalin helped me to solve the pysch error when installing ruby 3.2.1

**Comment by kassiansun** [comment]
On the other hand, I encountered another weird issue with `psych`, finally I realize that it was checking the `psych` gem from my ruby 3.2.0 installation.. Don't know why, but after cleared the old gem data, it installed successfully.

The build log says this:
```
Loading the Ruby yaml extension failed (incompatible library version - /Users/kassian/.local/share/gem/ruby/3.2.0/gems/psych-5.1.0/lib/psych.bundle)
ERROR: Ruby install aborted due to missing extensions
```

p.s. I was manually setting `GEM_HOME` to access `gem` path easily, but actually this name is used by gem, that's why `gem` was detecting installations in the 3.2.0 path.

**Comment by mnakayama1205** [comment]
I encountered the same error. 
After migrating intel homebrew to ARM homebrew, the error is completely disappeared.
